sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala

```diff
@@ -66,13 +76,151 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
+
+            if (nonChainSuffix.isEmpty) {
+              log.error("Could not recover storages inconsistency, could not find a rollback point in history")
+              None
+            } else {
+              val rollback_to = nonChainSuffix.get.head
+              val child_block = nonChainSuffix.get.tail.headOption
+              if (!child_block.isEmpty) {
+                log.info(s"Child ${BytesUtils.toHexString(idToBytes(child_block.get))} is in history")
+                log.info(s"Child info ${restoredHistory.blockInfoById(child_block.get)}")
+              }
+
+              // since the update order is state --> wallet --> history
+              // we can rollback both state and wallet to current best block in history or the ancestor of state block in active chain (which might as well be the same)
+              log.warn(s"Inconsistent storage and history, rolling back state and wallets to history best block id = ${rollback_to}")
+
+              val rolledBackState = restoredState.rollbackTo(idToVersion(rollback_to))
+              val rolledBackWallet = restoredWallet.rollback(idToVersion(rollback_to))
+
+              (rolledBackState, rolledBackWallet) match {
+                case (Success(s), Success(w)) =>
+                  log.debug("State and wallet succesfully rolled back")
+                  dumpStorages
+                  Some((restoredHistory, s, w, restoredMempool))
+                case (Failure(e), _) =>
+                  log.error("State roll back failed")
+                  context.system.eventStream.publish(RollbackFailed)
+                  None
+                case (_, Failure(e)) =>
+                  log.error("Wallet roll back failed")
+                  context.system.eventStream.publish(RollbackFailed)
+                  None
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  override def restoreState(): Option[(HIS, MS, VL, MP)] = {
+    log.info("Restoring persistent state from storage...")
+    val restoredData = for {
+      history <- SidechainHistory.restoreHistory(historyStorage, consensusDataStorage, params, semanticBlockValidators(params), historyBlockValidators(params))
+      state <- SidechainState.restoreState(stateStorage, forgerBoxStorage, utxoMerkleTreeStorage, params, applicationState)
+      wallet <- SidechainWallet.restoreWallet(sidechainSettings.wallet.seed.getBytes, walletBoxStorage, secretStorage,
+        walletTransactionStorage, forgingBoxesInfoStorage, cswDataStorage, params, applicationWallet)
+      pool <- Some(SidechainMemoryPool.emptyPool)
+    } yield (history, state, wallet, pool)
+
+    val result = checkAndRecoverStorages((restoredData))
+    result
+  }
+
+  def dumpStorages : Unit = {
+
+    try {
+      val m = getStorageVersions.map{ case(k, v) => {"%-36s".format(k) + ": " + v}}
+      m.foreach(x => log.debug(s"${x}"))
+      log.debug(s"    ForgingBoxesInfoStorage vers:    ${forgingBoxesInfoStorage.rollbackVersions.slice(0, 10)}")
```

Comment:
I reduced the number of versions to print (from 10 to 3) and reduced the level to 'trace'. Such storage can be exceptionally not aligned with other storage versions (it might have been updated upon consensus epoch switch even before the state update)